### PR TITLE
Disallow direct editing of EM_TOKEN or ROOT_TOKEN

### PR DESCRIPTION
**Pull request by aqkhan**
Referenced to #353

### CODE REVIEW

File: src/components/Editable.js [hunk 1]
```diff
@@ -69,7 +71,7 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
   const thoughtsResolved = contextChain.length ? chain(contextChain, thoughtsRanked) : thoughtsRanked
   const value = head(showContexts ? contextOf(thoughts) : thoughts) || ''
   const thoughtMeta = meta(thoughts)
-  const readonly = thoughtMeta.readonly
+  const readonly = (isRoot(thoughts) || isEM(thoughts)) ? true : thoughtMeta.readonly
```
Comment (raineorshine) [inline src/components/Editable.js]: You can simplify to `isRoot(thoughts) || isEM(thoughts) || thoughtMeta.readonly`

File: src/components/Thought.js [hunk 2]
```diff
@@ -151,6 +153,18 @@ export const Thought = connect(({ cursor, cursorBeforeEdit, expanded, expandedCo
 
       const { thoughtsRanked: thoughtsFrom } = monitor.getItem()
       const thoughtsTo = props.thoughtsRankedLive
+      const isRootOrEM = isRoot(thoughtsFrom) || isEM(thoughtsFrom)
+      const oldContext = rootedContextOf(thoughtsFrom)
+      const newContext = rootedContextOf(thoughtsTo)
+      const sameContext = equalArrays(oldContext, newContext)
```
Comment (raineorshine) [inline src/components/Thought.js]: Could you explain the use of `sameContext` here? I was thinking that we should error just on `isRootOrEM`, but let me know if I'm missing something.

**Comment by aqkhan** [comment]
I've halted the propagation of `existingThoughtChange` and `existingThoughtMove` in the source components, i.e `Thought` and `SubThought` components. That allows better handling of error invocation. Halting `existingThoughtDelete` can disable the deletion of parent node with backspace, let me know of that's desired. 
PS: The read-only items are still editable in UI, causing a number of issues.

File: src/components/Thought.js (hunk 2 above)
Comment (aqkhan) [reply src/components/Thought.js]: `sameContext ` was used to determine if the ROOT_TOKEN/EM_TOKEN is not being dropped into another context, i.e. dropped as a sub-thought.

**Comment by raineorshine** [comment]
> I've halted the propagation of `existingThoughtChange` and `existingThoughtMove` in the source components, i.e `Thought` and `SubThought` components. That allows better handling of error invocation.

Yes, but this doesn't solve the issue of the other commands that modify thoughts, such as `indent`, `outdent`, `subcategorizeOne`, `subcategorizeAll`, etc.

> Halting `existingThoughtDelete` can disable the deletion of parent node with backspace, let me know of that's desired.

`existingThoughtDelete` should be halted if the thought is `EM` or `ROOT`. I'm not sure I follow you regarding the parent node issue. Backspace deletes the current node not the parent node as far as I know.

File: src/components/Thought.js (hunk 2 above)
Comment (raineorshine) [reply src/components/Thought.js]: Okay, I understand now. Thanks!

**Comment by aqkhan** [comment]
> Yes, but this doesn't solve the issue of the other commands that modify thoughts, such as indent, outdent, subcategorizeOne, subcategorizeAll, etc.
Yes, that can only be controlled from the reducers. I will make those updates accordingly. 
> existingThoughtDelete should be halted if the thought is EM or ROOT. I'm not sure I follow you regarding the parent node issue. Backspace deletes the current node not the parent node as far as I know.
I was referring to a root node here. So, just to be sure, an `EM` or `ROOT` node should be deletable but not mutable?

**Comment by raineorshine** [comment]
> So, just to be sure, an `EM` or `ROOT` node should be deletable but not mutable?
The `EM_TOKEN` and `ROOT_TOKEN` nodes should not be deletable or mutable. You can add subthoughts to them though.

**Comment by aqkhan** [comment]
Raine, The following action-creators have now been halted on the nodes of EM_TOKEN and ROOT_TOKEN:
- [Indent](https://github.com/cybersemics/em/blob/dev/src/action-creators/indent.js)
- [Outdent/De-indent](https://github.com/cybersemics/em/blob/dev/src/action-creators/outdent.js)
- [SubCategorizeOne](https://github.com/cybersemics/em/blob/dev/src/action-creators/subCategorizeOne.js)
- [SubCategorizeAll](https://github.com/cybersemics/em/blob/dev/src/action-creators/subCategorizeAll.js)

Some actions are being halted on the account of the node being readonly. If readonly meta is removed by the user, those nodes become mutable. Additional checks can be added to avoid mutations on these nodes even if the readonly meta is removed. Please share your thoughts on this. 

Additionally the EM_TOKEN or ROOT_TOKEN can be removed entirely as of now, let me know if that needs to be disabled as well. 

PS: Additional checks can be put in the aforementioned reducers as well, but that would be redundant given the mutation has now being controlled from action-creators. I would like to have your input on this.

File: src/action-creators/outdent.js [hunk 3]
```diff
@@ -8,12 +11,32 @@ import {
...
+import { isEM } from '../util/isEM.js'
+import { isRoot } from '../util/isRoot.js'
 
 export const outdent = () => dispatch => {
   const { cursor } = store.getState()
   if (cursor && cursor.length > 1) {
 
+    // Cancel if a direct child of EM_TOKEN or ROOT_TOKEN
+    if (isEM(contextOf(cursor)) || isRoot(contextOf(cursor))) {
+      error(`Child of "${isEM(contextOf(cursor)) ? 'EM_TOKEN' : 'ROOT_TOKEN'}" may not be de-indented.`)
```
Comment (raineorshine) [inline src/action-creators/outdent.js]: User-facing text should use "Subthought" instead of "Child."

File: src/components/Editable.js [hunk 4]
```diff
@@ -128,6 +128,10 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
     }
 
     const oldValueClean = oldValue === EM_TOKEN ? 'em' : ellipsize(oldValue)
+    if (isEM(thoughts) || isRoot(thoughts)) {
+      error(`"${isEM(thoughts) ? 'EM_TOKEN' : 'ROOT_TOKEN'}" cannot be edited.`)
```
Comment (raineorshine) [inline src/components/Editable.js]: User-facing text should refer to `EM_TOKEN` as the "**em** context" and `ROOT_TOKEN` as the "home context".

File: src/util/rootContextOf.js [hunk 5]
```diff
@@ -0,0 +1,2 @@
+/** Gets the root context of a context */
+export const rootContextOf = context => context.slice(0, 1)
```
Comment (raineorshine) [inline src/util/rootContextOf.js]: Is this used any longer?

File: src/util.js [hunk 6]
```diff
@@ -15,6 +15,7 @@ import { componentToThought } from './util/componentToThought.js'
 import { conjunction } from './util/conjunction.js'
 import { contextChainToPath } from './util/contextChainToPath.js'
 import { contextOf } from './util/contextOf.js'
+import { rootContextOf } from './util/rootContextOf'
```
Comment (raineorshine) [inline src/util.js]: Note that other imports are alphabetically sorted.

File: src/util.js [hunk 7]
```diff
@@ -135,6 +136,7 @@ export {
   conjunction,
   contextChainToPath,
   contextOf,
+  rootContextOf,
```
Comment (raineorshine) [inline src/util.js]: Note that these keys are alphabetically sorted.

**Comment by raineorshine** [comment]
> Raine,
> The following action-creators have now been halted on the nodes of EM_TOKEN and ROOT_TOKEN:
> 
> * [Indent](https://github.com/cybersemics/em/blob/dev/src/action-creators/indent.js)
> * [Outdent/De-indent](https://github.com/cybersemics/em/blob/dev/src/action-creators/outdent.js)
> * [SubCategorizeOne](https://github.com/cybersemics/em/blob/dev/src/action-creators/subCategorizeOne.js)
> * [SubCategorizeAll](https://github.com/cybersemics/em/blob/dev/src/action-creators/subCategorizeAll.js)
> 

Thanks!

> Some actions are being halted on the account of the node being readonly. If readonly meta is removed by the user, those nodes become mutable. Additional checks can be added to avoid mutations on these nodes even if the readonly meta is removed. Please share your thoughts on this.

I am aware of this. For now I am going to leave it as-is, but in the future we may want to close that door. Good observation.

> Additionally the EM_TOKEN or ROOT_TOKEN can be removed entirely as of now, let me know if that needs to be disabled as well.

Yes, delete should be prevented on `EM_TOKEN` and `ROOT_TOKEN`

> Additional checks can be put in the aforementioned reducers as well, but that would be redundant given the mutation has now being controlled from action-creators. I would like to have your input on this.

I don't have a clear sense either way, so I will defer to you on this one.

File: src/util/rootContextOf.js (hunk 5 above)
Comment (aqkhan) [reply src/util/rootContextOf.js]: This has been removed.

**Comment by aqkhan** [comment]
Requested updates have been made.

File: src/action-creators/indent.js [hunk 8]
```diff
@@ -32,7 +32,7 @@ export const indent = () => dispatch => {
 
     // cancel if cursor is EM_TOKEN or ROOT_TOKEN
     if (isEM(cursor) || isRoot(cursor)) {
-      error(`"${isEM(cursor) ? 'EM_TOKEN' : 'ROOT_TOKEN'}" may not be indented.`)
+      error(`"${isEM(cursor) ? 'em context' : 'home context'}" may not be indented.`)
```
Comment (raineorshine) [inline src/action-creators/indent.js]:
```js
error(`The "${isEM(cursor) ? 'em' : 'home'}" context may not be indented.`)
```

File: src/action-creators/outdent.js [hunk 9]
```diff
@@ -24,7 +24,7 @@ export const outdent = () => dispatch => {
 
     // Cancel if a direct child of EM_TOKEN or ROOT_TOKEN
     if (isEM(contextOf(cursor)) || isRoot(contextOf(cursor))) {
-      error(`Child of "${isEM(contextOf(cursor)) ? 'EM_TOKEN' : 'ROOT_TOKEN'}" may not be de-indented.`)
+      error(`Subthought of "${isEM(contextOf(cursor)) ? 'em context' : 'home context'}" may not be de-indented.`)
```
Comment (raineorshine) [inline src/action-creators/outdent.js]: "Subthoughts of the..."

File: src/action-creators/subCategorizeAll.js [hunk 10]
```diff
@@ -31,7 +31,7 @@ export const subCategorizeAll = () => dispatch => {
 
   // Cancel if a direct child of EM_TOKEN or ROOT_TOKEN
   if (isEM(contextOf(cursor)) || isRoot(contextOf(cursor))) {
-    error(`Child of "${isEM(contextOf(cursor)) ? 'EM_TOKEN' : 'ROOT_TOKEN'}" may not be de-indented.`)
+    error(`Subthought of "${isEM(contextOf(cursor)) ? 'en context' : 'home context'}" may not be de-indented.`)
```
Comment (raineorshine) [inline src/action-creators/subCategorizeAll.js]: "Subthoughts of the..."

File: src/action-creators/subCategorizeOne.js [hunk 11]
```diff
@@ -28,7 +28,7 @@ export const subCategorizeOne = () => dispatch => {
 
   // Cancel if a direct child of EM_TOKEN or ROOT_TOKEN
   if (isEM(contextOf(cursor)) || isRoot(contextOf(cursor))) {
-    error(`Child of "${isEM(contextOf(cursor)) ? 'EM_TOKEN' : 'ROOT_TOKEN'}" may not be de-indented.`)
+    error(`Child of "${isEM(contextOf(cursor)) ? 'em context' : 'home context'}" may not be de-indented.`)
```
Comment (raineorshine) [inline src/action-creators/subCategorizeOne.js]: "Subthoughts of the..."

File: src/components/Editable.js [hunk 12]
```diff
@@ -129,7 +129,7 @@ export const Editable = connect()(({ isEditing, thoughtsRanked, contextChain, sh
 
     const oldValueClean = oldValue === EM_TOKEN ? 'em' : ellipsize(oldValue)
     if (isEM(thoughts) || isRoot(thoughts)) {
-      error(`"${isEM(thoughts) ? 'EM_TOKEN' : 'ROOT_TOKEN'}" cannot be edited.`)
+      error(`"${isEM(thoughts) ? 'em context' : 'home context'}" cannot be edited.`)
```
Comment (raineorshine) [inline src/components/Editable.js]: "The..."

File: src/components/Subthoughts.js [hunk 13]
```diff
@@ -140,7 +140,7 @@ export const Subthoughts = connect(({ cursorBeforeEdit, cursor, contextViews, th
         if (isRootOrEM && !sameContext) {
           store.dispatch({
             type: 'error',
-            value: `Cannot move ${isRoot(thoughtsFrom) ? `"root"` : `"em"`} to another context.`
+            value: `Cannot move ${isRoot(thoughtsFrom) ? `"home context"` : `"em context"`} to another context.`
```
Comment (raineorshine) [inline src/components/Subthoughts.js]: "Cannot move the..."

File: src/components/Thought.js [hunk 14]
```diff
@@ -161,7 +161,7 @@ export const Thought = connect(({ cursor, cursorBeforeEdit, expanded, expandedCo
       if (isRootOrEM && !sameContext) {
         store.dispatch({
           type: 'error',
-          value: `Cannot move ${isRoot(thoughtsFrom) ? `"root"` : `"em"`} to another context.`
+          value: `Cannot move ${isRoot(thoughtsFrom) ? `"home context"` : `"em context"`} to another context.`
```
Comment (raineorshine) [inline src/components/Thought.js]: "Cannot move the..."

File: src/shortcuts/delete.js [hunk 15]
```diff
@@ -9,13 +9,18 @@ import {
...
+  isRoot
 } from '../util.js'
 
 const exec = e => {
   const { cursor } = store.getState()
 
   if (cursor) {
-    if (meta(pathToContext(cursor)).readonly) {
+    if (isEM(cursor) || isRoot(cursor)) {
+      error(`"${isEM(cursor) ? 'em context' : 'home context'}" cannot be deleted.`)
```
Comment (raineorshine) [inline src/shortcuts/delete.js]: "The..."

**Comment by aqkhan** [comment]
Error messages have been updated in the [latest commit](https://github.com/aqkhan/em/commit/7031463c744c18b4703e1fd5060d6f9c82380d93).

**Comment by raineorshine** [comment]
Thanks!

**Pull request merged**